### PR TITLE
rcl_send_response returns RCL_RET_TIMEOUT.

### PR DESCRIPTION
**Pull request by fujitatomoya**
part of https://github.com/ros2/ros2/issues/1253

### CODE REVIEW

File: rcl/src/rcl/service.c [hunk 1]
```diff
@@ -373,16 +374,18 @@ rcl_send_response(
...
   RCL_CHECK_FOR_NULL_WITH_MSG(options, "Failed to get service options", return RCL_RET_ERROR);
 
-  if (rmw_send_response(
-      service->impl->rmw_handle, request_header, ros_response) != RMW_RET_OK)
-  {
+  ret = rmw_send_response(service->impl->rmw_handle, request_header, ros_response);
+  if (ret != RMW_RET_OK) {
     RCL_SET_ERROR_MSG(rmw_get_error_string().str);
+    if (ret == RMW_RET_TIMEOUT) {
+      return RCL_RET_TIMEOUT;
```
Comment (fujitatomoya) [inline rcl/src/rcl/service.c:381]: only pointing out `TIMEOUT` case, anything else can be reported as `ERROR` as it is implemented.

**Comment by fujitatomoya** [comment]
for both `rclcpp` and `rclpy` need to check the return code, and throws exception accordingly. we would need to add exception for timeout for client interface. Behavior in client libraries need to be discussed before implementation.

**Comment by fujitatomoya** [comment]
see https://github.com/ros2/rmw/pull/350#issuecomment-1477213924

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18288/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12814/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18955/)

**Review by audrow — APPROVED** [review]
Looks good to me!

**Comment by fujitatomoya** [comment]
@asorbini can i ask your help here?

so what we are trying to do here, if sending response times out, we will print the warning instead of generating exception for the client libraries, so that executor can proceed the next task.

as far as i see, `rmw_connextdds` does not return `RMW_RET_TIMEOUT` via `rmw_send_response`, it always returns `RMW_RET_ERROR` if anything wrong happens? (`rmw_fastrtps` and `rmw_cyclonedds` returns `RMW_RET_TIMEOUT`)

https://github.com/ros2/rmw_connextdds/blob/rolling/rmw_connextdds_common/src/ndds/dds_api_ndds.cpp#L709-L716

which i think that is fine because that is implementation detail, but just in case i wanted to share this with you since the same problem will exist with `rmw_connextdds` to generate the exception here.

just checking the behavior of `rmw_connextdds` to see if my understanding is correct.

thanks in advance,
Tomoya

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19079/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13606/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19825/)

**Review by clalancette — COMMENTED** [review]
The code here looks good to me, but likely needs a rebase to pass CI.

**Comment by fujitatomoya** [comment]
@clalancette thanks for the review, i just did rebase on rolling.

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19133/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13693/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19869/)

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19156/)

windows failure is unrelated.

**Review by clalancette — APPROVED** [review]

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19344/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13871/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20065/)

**Comment by fujitatomoya** [comment]
[Originally this issue](https://github.com/ros2/ros2/issues/1253) was reported with Galactic which is already E.O.L. Now we can see many related issues, i think it would be better to backport these fixes to Humble/Iron as well. and there is no ABI break.

**Comment by fujitatomoya** [comment]
@Mergifyio backport iron humble

**Comment by fujitatomoya** [comment]
lets give some soak time like a week for rolling before merging this backport PRs, then i can start the CI for PRs for Humble and Iron backport.

**Pull request merged**